### PR TITLE
Menu

### PR DESCRIPTION
**Pull request by jeremiehelme**

### CODE REVIEW

File: protected/widgets/menu/SideMenu.php [hunk 1]
```diff
@@ -20,15 +21,17 @@ class SideMenu extends Widget
     public $params = [];
     public $title;
     public $footer;
+    public $foldable = false;
 
     protected function registerClientScript()
     {
         $id = Json::encode($this->getId());
-
+        if($this->foldable == true) {
```
Comment (SamMousa) [inline protected/widgets/menu/SideMenu.php]:
```suggestion
        if($this->foldable) {
```

File: protected/widgets/menu/SideMenu.php [hunk 2]
```diff
@@ -20,15 +21,17 @@ class SideMenu extends Widget
...
         $js = <<<JS
+        
             document.getElementById($id).addEventListener('click', e =>  {
-                if (e.target.matches('header *')) {
```
Comment (SamMousa) [inline protected/widgets/menu/SideMenu.php]: Why did you remove this code?

File: protected/widgets/menu/SideMenu.php [hunk 3]
```diff
@@ -53,9 +57,14 @@ public function init()
         ];
 
         Html::addCssClass($options, 'menu');
+        if($this->foldable) Html::addCssClass($options, 'foldable');
         $this->registerClientScript();
         echo Html::beginTag('div', $options);
         echo Html::img("/img/HeRAMS.png");
+        if($this->foldable == true) {
```
Comment (SamMousa) [inline protected/widgets/menu/SideMenu.php]:
```suggestion
        if ($this->foldable) {
```

File: protected/widgets/menu/SideMenu.php [hunk 4]
```diff
@@ -53,9 +57,14 @@ public function init()
         ];
 
         Html::addCssClass($options, 'menu');
+        if($this->foldable) Html::addCssClass($options, 'foldable');
```
Comment (SamMousa) [inline protected/widgets/menu/SideMenu.php]: Don't use inline ifs, they reduce readability. (https://www.php-fig.org/psr/psr-2/)

```suggestion
    if ($this->foldable) {
        Html::addCssClass($options, 'foldable');
    }
```

File: protected/widgets/menu/assets/menu.css [hunk 5]
```diff
@@ -29,53 +29,60 @@
...
+  border-top-right-radius: 0px;
+  background-color: white;
+  grid-area: sidebar;
+  font-size: 1em;
+  z-index: 10000;
+  display: grid;
+  height: 100%;
+  grid-template-columns: 1fr;
+  grid-template-rows: auto auto auto 1fr auto;
+  grid-template-areas: "image" "title" "separator" "nav" "footer";
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: You should not randomly reformat code, it makes changes harder to find.
In the case of `grid-template-areas` reformatting makes it less readable; by using new lines you actually make sure that the value is an ascii art representation of your layout.

https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-areas

File: protected/widgets/menu/assets/menu.css [hunk 6]
```diff
@@ -29,53 +29,60 @@
 }
 
 .menu {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Don't change indenting, use the code style that you find.

File: composer.lock [hunk 7]
```diff
@@ -1,6361 +0,0 @@
-{
```
Comment (SamMousa) [inline composer.lock]: This file should not be part of the PR.

File: public/css/dashboard.css [hunk 8]
```diff
@@ -683,4 +732,8 @@ body[data-modal]:after {
     /*color: #92929b;*/
 }
 
+.footer {
```
Comment (SamMousa) [inline public/css/dashboard.css]: What is this for? If the footer needs to be hidden, why not make its rendering optional in the widget (just like the foldable feature).

Comment (jeremiehelme) [reply public/css/dashboard.css]: There is two footer, one in the menu, and another one just below. 
I'm not sure what the footer below is made for, and when / if we can remove it

Comment (SamMousa) [reply public/css/dashboard.css]: Can you show me a screenshot?

Comment (jeremiehelme) [reply public/css/dashboard.css]: footer

Comment (SamMousa) [reply public/css/dashboard.css]: Well yeah, but maybe it shouldn't be rendered in the first place; this looks like a bug to me ;-)
It definitely is. I'll fix the bug instead of working around it in css :)

Comment (SamMousa) [reply public/css/dashboard.css]: it is now fixed in master, you should be able to pull that change into your branch.

Comment (jeremiehelme) [reply public/css/dashboard.css]: Ok thanks, it works now. I removed the fix from css

File: protected/widgets/menu/assets/menu.css [hunk 9]
```diff
@@ -23,61 +23,68 @@
...
-            "separator"
-            "nav"
-            "footer"
+    grid-template-areas: "image" "title" "separator" "nav" "footer";
 }
-
-@keyframes fadein {
-    from { opacity: 0; }
-    to   { opacity: 1; }
+.menu h1 {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This selector is duplicated

File: protected/widgets/menu/assets/menu.css [hunk 10]
```diff
@@ -23,61 +23,68 @@
...
+    color: #4c4c4c;
+    text-transform: uppercase;
+    font-size: 18px;
+    letter-spacing: 8px;
+    margin: 0 auto;
+    text-align: center;
+    font-weight: 400;
+}
+.menu img {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This selector is duplicated

File: protected/widgets/menu/assets/menu.css [hunk 11]
```diff
@@ -23,61 +23,68 @@
...
 }
+
 .menu .footer .left, .menu .footer .right {
     display: inline;
     font-size: 48px;
-    position: absolute;;
+    position: absolute;
     padding: 0;
-    top: calc((100% - 48px - var(--status-height)) / 2);
+    top: 60px;
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This looks wrong to me; the footer of the menu should not be positioned in a fixed distance from the top, right?

File: protected/widgets/menu/assets/menu.css [hunk 12]
```diff
@@ -96,11 +103,88 @@
...
+    color: #bbb;
+    position: absolute;
+    top: 1px;
+    right: 10px;
+    font-size: 18px;
+    line-height: 18px;
+    margin-right: 10px;
+    margin-top: var(--item-padding-vertical);
+}
+.menu nav section header .icon.icon-chevron-right {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This looks like it would break if I decide to use a different icon.

File: protected/widgets/menu/assets/menu.css [hunk 13]
```diff
@@ -96,11 +103,88 @@
...
+    display: block;
+}
+.menu nav section header .icon.icon-chevron-down {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This looks like it would break if I decide to use a different icon.

File: protected/widgets/menu/assets/menu.css [hunk 14]
```diff
@@ -96,11 +103,88 @@
...
+    display: none;
+}
+.menu nav section header.expanded .icon.icon-chevron-right {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This looks like it would break if I decide to use a different icon.

File: protected/widgets/menu/assets/menu.css [hunk 15]
```diff
@@ -96,11 +103,88 @@
...
+    display: none;
+}
+.menu nav section header.expanded .icon.icon-chevron-down {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This looks like it would break if I decide to use a different icon.

File: protected/widgets/menu/assets/menu.css [hunk 16]
```diff
@@ -23,61 +23,68 @@
     padding-left: 2em;
 }
 
-
 .menu ul {
     margin: 0;
 }
 
 .menu {
-    overflow: hidden;
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Why remove this line

File: protected/widgets/menu/assets/menu.css [hunk 17]
```diff
@@ -23,61 +23,68 @@
...
-    --animation-duration: 500ms;
     --indicator-width: 3px;
     --indent: 20px;
-    --item-padding-vertical: 10px;
+    --item-padding-vertical: 14px;
+    overflow: hidden;
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: And re-add it here?

File: protected/widgets/menu/assets/menu.css [hunk 18]
```diff
@@ -23,61 +23,68 @@
...
-    grid-template-areas:
-            "image"
-            "title"
-            "separator"
-            "nav"
-            "footer"
+    grid-template-areas: "image" "title" "separator" "nav" "footer";
 }
-
-@keyframes fadein {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Why is this no longer needed?

File: protected/widgets/menu/assets/menu.css [hunk 19]
```diff
@@ -181,17 +263,19 @@
...
 
-.menu .footer .stats > div {
+.menu .footer .stats>div {
     display: none;
 }
-.menu .footer .stats > div:first-child {
+
+.menu .footer .stats>div:first-child {
     display: block;
     animation: fadein 2s;
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Isn't this where we need the keyframes?

File: protected/widgets/menu/assets/menu.css [hunk 20]
```diff
@@ -209,15 +293,51 @@
...
         max-width: 170px;
-        margin-top: 10px;
-        margin-bottom: 10px;
-        margin-left: auto;
-        margin-right: auto;
+    }
+    .menu.foldable .toggleMenu {
+        display: none;
+    }
+    .menu.foldable.expanded .toggleMenu.icon.icon-chevron-right {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Brittle; refers to specific icon

File: protected/widgets/menu/assets/menu.css [hunk 21]
```diff
@@ -209,15 +293,51 @@
...
+        display: none;
+    }
+    .menu.foldable.expanded .toggleMenu.icon.icon-chevron-down {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Brittle; refers to specific icon

File: protected/widgets/menu/assets/menu.css [hunk 22]
```diff
@@ -209,15 +293,51 @@
...
+        display: block;
+    }
+    .menu.foldable .toggleMenu.icon.icon-chevron-right {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Brittle; refers to specific icon

File: protected/widgets/menu/assets/menu.css [hunk 23]
```diff
@@ -209,15 +293,51 @@
...
+        display: block;
+    }
+    .menu.foldable .toggleMenu.icon.icon-chevron-down {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: Brittle; refers to specific icon

File: public/css/dashboard.css [hunk 24]
```diff
@@ -438,12 +493,6 @@ body {
                 "main main";
     }
 
-    body.no-title {
```
Comment (SamMousa) [inline public/css/dashboard.css]: Why remove this

File: public/css/dashboard.css [hunk 25]
```diff
@@ -390,6 +390,61 @@ body {
...
+      "sidebar sidebar"
+      "filters filters"
+      "main main";*/
+    grid-template-areas:
+      "sidebar title title user"
+      "sidebar filters filters filters"
+      "sidebar main main main";
+  }
+
+  body.no-title {
```
Comment (SamMousa) [inline public/css/dashboard.css]: Why add this?

File: protected/widgets/menu/assets/menu.css (hunk 11 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: .left and .right are the arrows in the footer. 
You are right though, they should be auto-centered vertically, if the footer height change

File: protected/widgets/menu/assets/menu.css [hunk 26]
```diff
@@ -1,15 +1,6 @@
 /**
  * This CSS file is menu widget
  */
-.menu h1 {
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This block is removed here and added again later.

File: public/css/dashboard.css (hunk 25 above)
Comment (jeremiehelme) [reply public/css/dashboard.css]: I've moved body.no-title closer to body declaration, it makes more sense to me as it helps to see the difference in the grid

File: public/css/dashboard.css (hunk 24 above)
Comment (jeremiehelme) [reply public/css/dashboard.css]: I've moved body.no-title closer to body declaration, it makes more sense to me as it helps to see the difference in the grid

File: protected/widgets/menu/assets/menu.css (hunk 19 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: Yes, I've put it back, sorry

File: protected/widgets/menu/assets/menu.css (hunk 26 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: It was duplicated, I've regrouped the two

File: protected/widgets/menu/assets/menu.css (hunk 16 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: I though it would be better to have the variables declaration at the beginning. Didn't think it would appear in the commit. I'll try not to do that again

File: public/css/dashboard.css (hunk 24 above)
Comment (SamMousa) [reply public/css/dashboard.css]: Don't do that in this PR, separate functional changes from reformatting in a separate PR.

It's fine with me if you want to reorganize the whole file but don't do that in a PR that also changes in the contents functionally.

For now this is okay; but it just takes me a whole lot longer to identify what's going on.

--> side note, `.no-title` view is broken for the dashboard; but that's a different issue.

File: protected/widgets/menu/assets/menu.css [hunk 27]
```diff
@@ -23,61 +14,73 @@
...
 }
 
-
 .menu ul {
     margin: 0;
 }
 
 .menu {
-    overflow: hidden;
-    --animation-duration: 500ms;
```
Comment (SamMousa) [inline protected/widgets/menu/assets/menu.css]: This variable is needed, why was it removed?

File: protected/widgets/menu/assets/menu.css (hunk 11 above)
Comment (SamMousa) [reply protected/widgets/menu/assets/menu.css]: This still needs fixing.

File: public/css/dashboard.css (hunk 24 above)
Comment (jeremiehelme) [reply public/css/dashboard.css]: Ok, I understand, thanks

File: protected/widgets/menu/assets/menu.css (hunk 27 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: I've put it back

File: protected/widgets/menu/assets/menu.css (hunk 11 above)
Comment (jeremiehelme) [reply protected/widgets/menu/assets/menu.css]: I've put back the calc
